src/ring_theory/is_adjoin_root.lean

```diff
@@ -0,0 +1,578 @@
+
+/-- `repr` preserves addition, up to multiples of `f` -/
+lemma repr_add_sub_repr_add_repr_mem_span (h : is_adjoin_root S f) (x y : S) :
+  h.repr (x + y) - (h.repr x + h.repr y) ∈ ideal.span ({f} : set R[X]) :=
+by rw [← h.ker_map, ring_hom.mem_ker, map_sub, h.map_repr, map_add, h.map_repr, h.map_repr,
+       sub_self]
+
+/-- Extensionality of the `is_adjoin_root` structure itself. See `is_adjoin_root_monic.ext_elem`
+for extensionality of the ring elements. -/
+lemma ext_map (h h' : is_adjoin_root S f) (eq : ∀ x, h.map x = h'.map x) : h = h' :=
+begin
+  cases h, cases h', congr,
+  exact ring_hom.ext eq
+end
+
+/-- Extensionality of the `is_adjoin_root` structure itself. See `is_adjoin_root_monic.ext_elem`
+for extensionality of the ring elements. -/
+@[ext]
+lemma ext (h h' : is_adjoin_root S f) (eq : h.root = h'.root) : h = h' :=
+h.ext_map h' (λ x, by rw [← h.aeval_eq, ← h'.aeval_eq, eq])
+
+section lift
+
+variables {T : Type*} [comm_ring T] {i : R →+* T} {x : T} (hx : f.eval₂ i x = 0)
+
+include hx
+
+/-- Auxiliary lemma for `is_adjoin_root.lift` -/
+lemma eval₂_repr_eq_eval₂_of_map_eq (h : is_adjoin_root S f) (z : S) (w : R[X])
+  (hzw : h.map w = z) :
+  (h.repr z).eval₂ i x = w.eval₂ i x :=
+begin
+  rw [eq_comm, ← sub_eq_zero, ← h.map_repr z, ← map_sub, h.map_eq_zero_iff] at hzw,
+  obtain ⟨y, hy⟩ := hzw,
+  rw [← sub_eq_zero, ← eval₂_sub, hy, eval₂_mul, hx, zero_mul]
+end
+
+variables (i x) -- To match `adjoin_root.lift`
+
+/-- Lift a ring homomorphism `R →+* T` to `S →+* T` by specifying a root `x` of `f` in `T`,
+where `S` is given by adjoining a root of `f` to `R`. -/
+def lift (h : is_adjoin_root S f) : S →+* T :=
+{ to_fun := λ z, (h.repr z).eval₂ i x,
+  map_zero' := by rw [h.eval₂_repr_eq_eval₂_of_map_eq hx _ _ (map_zero _), eval₂_zero],
+  map_add' := λ z w, begin
+    rw [h.eval₂_repr_eq_eval₂_of_map_eq hx _ (h.repr z + h.repr w), eval₂_add],
+    { rw [map_add, map_repr, map_repr] }
+  end,
+  map_one' := by rw [h.eval₂_repr_eq_eval₂_of_map_eq hx _ _ (map_one _), eval₂_one],
+  map_mul' := λ z w, begin
+    rw [h.eval₂_repr_eq_eval₂_of_map_eq hx _ (h.repr z * h.repr w), eval₂_mul],
+    { rw [map_mul, map_repr, map_repr] }
+  end }
+
+variables {i x}
+
+@[simp] lemma lift_map (h : is_adjoin_root S f) (z : R[X]) :
+  h.lift i x hx (h.map z) = z.eval₂ i x :=
+by rw [lift, ring_hom.coe_mk, h.eval₂_repr_eq_eval₂_of_map_eq hx _ _ rfl]
+
+@[simp] lemma lift_root (h : is_adjoin_root S f) :
+  h.lift i x hx h.root = x :=
+by rw [← h.map_X, lift_map, eval₂_X]
+
+@[simp] lemma lift_algebra_map (h : is_adjoin_root S f) (a : R) :
+  h.lift i x hx (algebra_map R S a) = i a :=
+by rw [h.algebra_map_apply, lift_map, eval₂_C]
+
+variables [algebra R T] (hx' : aeval x f = 0)
+
+omit hx
+
+variables (x) -- To match `adjoin_root.lift_hom`
+
+/-- Lift the algebra map `R → T` to `S →ₐ[R] T` by specifying a root `x` of `f` in `T`,
+where `S` is given by adjoining a root of `f` to `R`. -/
+def lift_hom (h : is_adjoin_root S f) : S →ₐ[R] T :=
```

Comment:
Can you add some unicity of this lift? It is the only `alg_hom` that maps `r.root` to `x`.